crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,790 @@
+                            transform_to_filetime(meta.atime()),
+                            transform_to_filetime(meta.mtime()),
+                            i64::try_from(meta.size()).unwrap(),
+                            file_attribute,
+                            file_name.to_owned(),
+                        );
+                        let info2 = FileInformationClass::BothDirectory(info);
+                        Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                            ClientDriveQueryDirectoryResponse {
+                                device_io_reply: DeviceIoResponse::new(device_io_request, NtStatus::SUCCESS),
+                                buffer: Some(info2),
+                            },
+                        ))])
+                    } else {
+                        warn!("unsupported file class for query directory");
+                        Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                            ClientDriveQueryDirectoryResponse {
+                                device_io_reply: DeviceIoResponse::new(device_io_request, NtStatus::NOT_SUPPORTED),
+                                buffer: None,
+                            },
+                        ))])
+                    }
+                }
+                Err(error) => {
+                    warn!(%error, "Get metadata error");
+                    Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                        ClientDriveQueryDirectoryResponse {
+                            device_io_reply: DeviceIoResponse::new(device_io_request, not_found_status),
+                            buffer: None,
+                        },
+                    ))])
+                }
+            }
+        }
+    }
+}
+pub(crate) fn query_directory(
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: ServerDriveQueryDirectoryRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    match backend.file_path_map.get(&req_inner.device_io_request.file_id) {
+        Some(parent_pos_for_next) => {
+            let mut find_file_name = None;
+            if req_inner.initial_query > 0 {
+                if req_inner.path.ends_with('*') {
+                    let mut parent = backend.file_base.clone();
+                    let query_path = req_inner.path.replace('\\', "/");
+                    let len = query_path.len();
+                    // path ends with *, so its len > 0
+                    #[allow(clippy::arithmetic_side_effects)]
+                    parent.push_str(&query_path[0..len - 1]);
+                    let dirp = if let Ok(real_parent_c) = CString::new(parent.clone()) {
+                        // SAFETY: it is safe to call this on *nix platform
+                        unsafe { opendir(real_parent_c.as_ptr()) }
+                    } else {
+                        std::ptr::null::<DIR>().cast_mut()
+                    };
```

Comment:
suggestion: I suggest to encapsulate this logic in `BoxDir`, in a `new` function.

```rust
impl BoxDir {
    fn new(path: &Path) -> Result<Self> {
        /* … */    
    }
}
```